### PR TITLE
[StackedAreaPlotTests] - Revamp

### PR DESCRIPTION
**Pull request by bluong**
Trying to scope down on each of the uses of the derived functions on `StackedAreaPlot`

Test screenshot:
screen shot 2015-10-26 at 2 38 04 pm

### CODE REVIEW

**Comment by ztsai** [comment]
can you resolve the merge conflicts?
also, is there any reason why you are not using `assertLinePathEqualToDataPoints` and write a new method instead?

**Comment by bluong** [comment]
`assertLinePathEqualToDataPoints` is not enough, unless I create a data object that contains the stack offsets instead of the data itself.  Would the be preferred?

**Comment by ztsai** [comment]
I see
It seems to me there's some duplicated functionality in `assertLinePathEqualToDataPoints` and `areaVertices`
does it make sense for me to remove `assertLinePathEqualToDataPoints` and use `areaVertices` in the tests for LinePlot and AreaPlot?

File: test/plots/stackedAreaPlotTests.ts [hunk 1]
```diff
@@ -2,485 +2,346 @@
...
-      let xAxis = new Plottable.Axes.Numeric(xScale, "bottom");
-      new Plottable.Components.Table([[renderer], [xAxis]]).renderTo(svg);
+    describe("setting the x property", () => {
+      let svg: d3.Selection<void>;
+      let stackedAreaPlot: Plottable.Plots.StackedArea<number>;
+
+      beforeEach(() => {
+        svg = TestMethods.generateSVG();
+        let data1 = [
+          {foo: 1},
```
Comment (ztsai) [inline test/plots/stackedAreaPlotTests.ts]: probably just personal preference, but `foo`, `bar` are not good variable names

File: test/plots/stackedAreaPlotTests.ts [hunk 2]
```diff
@@ -2,485 +2,346 @@
...
+        let stackedAreas = stackedAreaPlot.content().selectAll("path");
+        assert.strictEqual(stackedAreas.size(),
+          stackedAreaPlot.datasets().length, "same number of area selections as datasets");
+
+        stackedAreas.each(function(d, i) {
+          let stackedAreaSelection = d3.select(this);
+          let areaVertices = TestMethods.areaVertices(stackedAreaSelection);
+          let areaXs = areaVertices.map((areaVertex) => areaVertex.x).slice(0, -2);
+          d.forEach((datum: any, index: number) => {
+            assert.closeTo(areaXs[index], constantValue, window.Pixel_CloseTo_Requirement, "cools");
```
Comment (ztsai) [inline test/plots/stackedAreaPlotTests.ts]: message

File: test/plots/stackedAreaPlotTests.ts (hunk 1 above)
Comment (bluong) [inline test/plots/stackedAreaPlotTests.ts]: These are attributes however.  What do you think we should rename them to be?

Comment (ztsai) [inline test/plots/stackedAreaPlotTests.ts]: from the usage, `foo` could be `x` and `bar` could be `offset`, but then again, it's just personal preference as there's not a clear mapping

File: test/plots/stackedAreaPlotTests.ts [hunk 3]
```diff
@@ -2,485 +2,346 @@
...
+        assert.strictEqual(stackedAreas.size(),
+          stackedAreaPlot.datasets().length, "same number of area selections as datasets");
+
+        stackedAreas.each(function(data, i) {
+          let stackedAreaSelection = d3.select(this);
+          let areaVertices = TestMethods.areaVertices(stackedAreaSelection);
+          let areaXs = areaVertices.map((areaVertex) => areaVertex.x).slice(0, -2);
+          data.forEach((datum: any, datumIndex: number) => {
+            let x = accessor(datum, datumIndex, stackedAreaPlot.datasets()[i]);
+            assert.closeTo(areaXs[datumIndex], x, window.Pixel_CloseTo_Requirement, "cools");
```
Comment (ztsai) [inline test/plots/stackedAreaPlotTests.ts]: message

File: test/plots/stackedAreaPlotTests.ts [hunk 4]
```diff
@@ -2,485 +2,346 @@
...
+        stackedAreaPlot.renderTo(svg);
+
+        let stackedAreas = stackedAreaPlot.content().selectAll("path");
+        assert.strictEqual(stackedAreas.size(),
+          stackedAreaPlot.datasets().length, "same number of area selections as datasets");
+
+        stackedAreas.each(function(d, i) {
+          let stackedAreaSelection = d3.select(this);
+          let areaVertices = TestMethods.areaVertices(stackedAreaSelection);
+          let areaXs = areaVertices.map((areaVertex) => areaVertex.x).slice(0, -2);
```
Comment (ztsai) [inline test/plots/stackedAreaPlotTests.ts]: can you check the last 2 points as well?

File: test/plots/stackedAreaPlotTests.ts [hunk 5]
```diff
@@ -2,485 +2,346 @@
...
+      });
+
+      it("can set the y0 property accessor to a constant value", () => {
+        let constantValue = 10;
+        stackedAreaPlot.y0(constantValue);
+        assert.strictEqual(stackedAreaPlot.y0().accessor(null, 0, null), constantValue, "property set to constant value");
+        stackedAreaPlot.destroy();
+      });
+
+      it("can set the y0 property accessor to be dependent on the input data", () => {
```
Comment (ztsai) [inline test/plots/stackedAreaPlotTests.ts]: can you also checked the rendered points?

File: test/plots/stackedAreaPlotTests.ts [hunk 6]
```diff
@@ -2,485 +2,346 @@
...
-      svg.remove();
+    describe("rendering on edge case scenarios", () => {
+      it("renders nothing when no data", () => {
+        let svg = TestMethods.generateSVG();
+        let stackedAreaPlot = new Plottable.Plots.StackedArea();
+        stackedAreaPlot.x(() => null);
+        // HACKHACK https://github.com/palantir/plottable/issues/2712 y scale must be set.
+        stackedAreaPlot.y(() => null, new Plottable.Scales.Linear());
+        stackedAreaPlot.renderTo(svg);
+        assert.strictEqual(stackedAreaPlot.selections().size(), 0, "test");
```
Comment (ztsai) [inline test/plots/stackedAreaPlotTests.ts]: message

File: test/plots/stackedAreaPlotTests.ts [hunk 7]
```diff
@@ -2,485 +2,346 @@
...
+        stackedAreaPlot.destroy();
+        svg.remove();
+      });
+
+      it("0 as a string coerces correctly and is not subject to off by one errors", () => {
```
Comment (ztsai) [inline test/plots/stackedAreaPlotTests.ts]: can you rewrite the message so that it reads semantically along with `it`?

Comment (ztsai) [inline test/plots/stackedAreaPlotTests.ts]: also, I believe we usually consider number string as invalid input in other plots
"0" is an edge case, can you test against non-zero number string and see it's skipped?

File: test/plots/stackedAreaPlotTests.ts [hunk 8]
```diff
@@ -2,485 +2,346 @@
...
+        stackedAreaPlot.destroy();
+      });
     });
 
-    it("disables downsampling", () => {
-      assert.strictEqual(renderer.downsamplingEnabled(), false, "downsampling is disabled by default");
-      renderer.downsamplingEnabled(true);
-      assert.strictEqual(renderer.downsamplingEnabled(), false, "downsampling will not be enabled by user");
-      svg.remove();
+    describe("rendering on edge case scenarios", () => {
```
Comment (ztsai) [inline test/plots/stackedAreaPlotTests.ts]: can you also test if x values of datasets don't align?

File: test/plots/stackedAreaPlotTests.ts (hunk 5 above)
Comment (bluong) [inline test/plots/stackedAreaPlotTests.ts]: I believe there is a bug here
https://github.com/palantir/plottable/issues/2861

**Comment by bluong** [comment]
I am unsure about using `assert` vs the `areaVertices` function.  My understanding is that I did not want to necessarily assert but only extract the relevant pieces of information and then continue with my test flow.

File: test/plots/stackedAreaPlotTests.ts (hunk 4 above)
Comment (bluong) [inline test/plots/stackedAreaPlotTests.ts]: Why so?

File: test/plots/stackedAreaPlotTests.ts (hunk 8 above)
Comment (bluong) [inline test/plots/stackedAreaPlotTests.ts]: `stackedAreaPlot` does not support this scenario.  How do you want me to test this?

Comment (ztsai) [inline test/plots/stackedAreaPlotTests.ts]: is this documented somewhere?

Comment (bluong) [inline test/plots/stackedAreaPlotTests.ts]: Well it's just that `stackedAreaPlot` will operate in a way that we would not recommend users to use.  If someone can come up with a detailed spec on how we want to do this, then we can align to that idea, but right now that spec is very vague

File: test/plots/stackedAreaPlotTests.ts (hunk 5 above)
Comment (ztsai) [inline test/plots/stackedAreaPlotTests.ts]: what about adding a test and then skip it?

File: test/plots/stackedAreaPlotTests.ts (hunk 4 above)
Comment (ztsai) [inline test/plots/stackedAreaPlotTests.ts]: for correctness of the endpoints?

Comment (bluong) [inline test/plots/stackedAreaPlotTests.ts]: I'll see how hard / useful it is.

File: test/plots/stackedAreaPlotTests.ts (hunk 5 above)
Comment (bluong) [inline test/plots/stackedAreaPlotTests.ts]: I personally don't see too much immediate value in adding a skipped test.  I would much rather create an issue for the bug and then make the test at that point.

File: test/plots/stackedAreaPlotTests.ts [hunk 9]
```diff
@@ -1,486 +1,373 @@
...
+        let stackedAreaSelection2 = d3.select(stackedAreaSelections[0][2]);
+
+        let areaVertices0 = TestMethods.areaVertices(stackedAreaSelection0);
+        let areaYs0 = areaVertices0.map((areaVertex) => areaVertex.y).slice(0, -2);
+        let areaVertices1 = TestMethods.areaVertices(stackedAreaSelection1);
+        let areaYs1 = areaVertices1.map((areaVertex) => areaVertex.y).slice(0, -2);
+        let areaVertices2 = TestMethods.areaVertices(stackedAreaSelection2);
+        let areaYs2 = areaVertices2.map((areaVertex) => areaVertex.y).slice(0, -2);
+
+        assert.closeTo(areaYs0[0], yScale.scale(<any> data0[0].y), window.Pixel_CloseTo_Requirement,
```
Comment (ztsai) [inline test/plots/stackedAreaPlotTests.ts]: why do you need to any cast data0?

File: test/plots/stackedAreaPlotTests.ts [hunk 10]
```diff
@@ -1,486 +1,373 @@
...
+          "dataset0 should have no offset");
+        assert.closeTo(areaYs1[0], yScale.scale(<any> data0[0].y), window.Pixel_CloseTo_Requirement,
+          "dataset1 should be offset with dataset0");
+        assert.closeTo(areaYs2[0], yScale.scale(data2[0].y + parseInt(<any> data0[0].y, 10)), window.Pixel_CloseTo_Requirement,
+          "dataset2 should be offset with only dataset0");
```
Comment (ztsai) [inline test/plots/stackedAreaPlotTests.ts]: this error message is misleading.  it's not offset with dataset1 because the value happens to be "0"

File: test/plots/stackedAreaPlotTests.ts (hunk 9 above)
Comment (bluong) [inline test/plots/stackedAreaPlotTests.ts]: `data0[0]` is a string and our current behavior is such that there is a coercion done when stacking from strings to numbers.  In accordance there must be a coercion done from `string`s to `number`s for the scale

File: test/plots/stackedAreaPlotTests.ts [hunk 11]
```diff
@@ -1,486 +1,373 @@
...
+        let areaVertices1 = TestMethods.areaVertices(stackedAreaSelection1);
+        let areaYs1 = areaVertices1.map((areaVertex) => areaVertex.y).slice(0, -2);
+        let areaVertices2 = TestMethods.areaVertices(stackedAreaSelection2);
+        let areaYs2 = areaVertices2.map((areaVertex) => areaVertex.y).slice(0, -2);
+
+        assert.closeTo(areaYs0[0], yScale.scale(<any> data0[0].y), window.Pixel_CloseTo_Requirement,
+          "dataset0 should have no offset");
+        assert.closeTo(areaYs1[0], yScale.scale(<any> data0[0].y), window.Pixel_CloseTo_Requirement,
+          "dataset1 should be offset with previous datasets");
+        assert.closeTo(areaYs2[0], yScale.scale(data2[0].y + parseInt(<any> data0[0].y, 10)), window.Pixel_CloseTo_Requirement,
```
Comment (ztsai) [inline test/plots/stackedAreaPlotTests.ts]: should be data2[0].y + data1[0].y + data1[0].y ?
